### PR TITLE
RegexPatternDetector detects implicit object creation

### LINKED ISSUE #47743: Target-typed Regex initialization doesn't syntax highlight the regular expression
**Version Used**: 
Version 16.8.0 Preview 3.0

**Steps to Reproduce**:

        public static readonly Regex ResistanceImageRegex = new(@"^\$\((.*)ResistanceImage\)$");

**Expected Behavior**:

I would expect it to be syntax highlighted like it is when I use the explicit constructor, instead of relying on target typed construction:

image

**Actual Behavior**:

As seen above, target-typing doesn't syntax highlight the regular expression.

### PR DESCRIPTION
**Pull request by meziantou**
fix #47743

### CODE REVIEW

File: src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs [hunk 1]
```diff
@@ -283,6 +283,16 @@ public bool IsRegexPattern(SyntaxToken token, CancellationToken cancellationToke
                     }
                 }
             }
+            else if (syntaxFacts.IsImplicitObjectCreationExpression(invocationOrCreation))
+            {
+                var constructor = _semanticModel.GetSymbolInfo(invocationOrCreation, cancellationToken).GetAnySymbol();
```
Comment (meziantou) [inline src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs:288]: In the case of an object creation (`new Regex()`), there is a quick check on the type name before querying the semantic model.
Do you think it is useful to try to detect the type name here before going to the semantic model? For instance, we could try to get the type from the left side of the assignation (`Regex r = new("")`). If so, which pattern should we support?

File: src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs [hunk 2]
```diff
@@ -1,4 +1,4 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
+// Licensed to the .NET Foundation under one or more agreements.
```
Comment (Youssef1313) [inline src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs:1]: This seems to *incorrectly* remove the BOM character at the beginning of the file.

File: src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionTests.cs [hunk 3]
```diff
@@ -172,6 +172,27 @@ void M()
             AssertJsonEquals(expected, results.Items.First());
         }
 
+        [Fact]
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/ProtocolUnitTests/Completion/CompletionTests.cs:175]: Add WorkItem attribute please :-)

File: src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs (hunk 1 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs:288]: Unfortunately, I can't see a way that would help. Implicit object creation could show up anywhere (like `Foo(new(""))`. So we're going to need semantics no matter what.

**Review by CyrusNajmabadi — APPROVED** [review]

File: src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs (hunk 2 above)
Comment (meziantou) [reply src/Features/Core/Portable/EmbeddedLanguages/RegularExpressions/LanguageServices/RegexPatternDetector.cs]: I don't know why I cannot accept your suggestion (buttons are disabled). I've fixed it manually.

**Review by CyrusNajmabadi — APPROVED** [review]
Thanks!

**Comment by CyrusNajmabadi** [comment]
Thanks for the contribution @meziantou !

**Pull request merged**